ghcide/src/Development/IDE/Plugin/Completions.hs

```diff
@@ -77,10 +79,19 @@ produceCompletions = do
               let env = hscEnv sess
               -- We do this to be able to provide completions of items that are not restricted to the explicit list
               res <- liftIO $ tcRnImportDecls env (dropListFromImportDecl <$> imps)
-              case res of
-                  (_, Just rdrEnv) -> do
+              inScope <- liftIO $ tcRnImportDecls env imps
+              case (res, inScope) of
+                  ((_, Just globalEnv), (_, Just inScopeEnv)) -> do
                       let uri = fromNormalizedUri $ normalizedFilePathToUri file
-                      cdata <- liftIO $ cacheDataProducer uri env (ms_mod ms) rdrEnv imps parsedDeps
+                          globalElts = globalRdrEnvElts globalEnv
+                          inScopeElts = globalRdrEnvElts inScopeEnv
+                          diff = getDiff globalElts inScopeElts
+                          go (First x : xs) ys = go xs $ (x, False) : ys
+                          go (Second x : xs) ys = go xs $ (x, True) : ys
+                          go (Both x _ : xs) ys = go xs $ (x, True) : ys
+                          go [] ys = ys
+                          rdrElts = go diff []
```

Comment:
do we need to reverse the `diff` list? If not, why don't you use `map` instead?

Reply:
Thanks, `map` should work here. I switched to `map`, which means the list is reversed back to positive order, and then I found the tests failed. This reveals a bug in code that we should not use `Eq` instance of `GlobalRdrElt`, because it compares not only `gre_name`, but `gre_imp` as well. However we don't want to compare this, since a same element in `globalEnv` and `inScopeEnv` may not have the same `gre_imp` (they are brought into scope in different ways), for example:

```
ZeroPad: fst | ZeroPad parent:FormatAdjustment
  imported from ‘Text.Printf’ at /run/user/1000/extra-dir-46668153349363/A.hs:2:1-42
ZeroPad: snd | ZeroPad parent:FormatAdjustment
  imported from ‘Text.Printf’ at /run/user/1000/extra-dir-46668153349363/A.hs:2:21-41
```

So it becomes clear that the order of the list affects the result.